shared/metrics/metrics.go

```diff
@@ -55,117 +65,211 @@ func FromContext(ctx context.Context) Metrics {
 var _ Metrics = (*PromMetrics)(nil)
 
 type PromMetrics struct {
-	gasBalance                    metrics.Gauge
```

Comment:
this was a bit more involved to implement, so if we feel like we really want to add this as a metric i can do in a separate PR